virt/xen/xen.py

```diff
@@ -0,0 +1,294 @@
+    # events that have occurred, you probably want to use this the first time you use event_from
+    token_from =''
+
+    # Register for events on all classes
+    event_types = ["*"]
+
+    def __init__(self, logger, config):
+        super(Xen, self).__init__(logger, config)
+        self.url = config.server
+        self.username = config.username
+        self.password = config.password
+        self.config = config
+
+        # Url must contain protocol (usualy https://)
+        if "://" not in self.url:
+            self.url = "https://%s" % self.url
+
+        self.filter = None
+
+    def _prepare(self):
+        """ Prepare for obtaining information from Xen server. """
+        self.logger.debug("Log into XEN pools %s" % self.url)
+        self.login()
+
+    def login(self):
+        # Login to server using given credentials
+        try:
+            # Don't log message containing password
+            self.session = XenAPI.Session(self.url)
+            self.session.xenapi.login_with_password(self.username, self.password)
+            self.logger.debug("I'm in XEN pools! with user %s" % self.username)
+
+        except NewMaster as nm:
+            self.logger.debug("Tengo un nuevo master?")
+            try:
+                self.session = XenAPI.Session('http://%s' % nm.new_master())
+                self.session.xenapi.login_with_password(self.username, self.password)
+                self.logger.debug("I'm in XEN pools with new master %s! with user %s" % (nm.new_master(), self.username))
+
+            except:
+                self.logger.exception("Unable to login to XENserver %s" % self.url)
+                raise
+
+        except:
+            self.logger.exception("Unable to login to XENserver %s" % self.url)
+            raise
+
+    def getHostGuestMapping(self):
+        hosts= self.session.xenapi.host.get_all()
+
+        mapping = {}
+
+        for host in hosts:
+
+            record= self.session.xenapi.host.get_record(host)
+            guests=[]
+
+            for resident in self.session.xenapi.host.get_resident_VMs(host):
+                vm= self.session.xenapi.VM.get_record(resident)
+
+                if vm['power_state'] == 'Running':
+                    state = virt.Guest.STATE_RUNNING
+                elif vm['power_state'] == 'Suspended':
+                    state = virt.Guest.STATE_PAUSED
+                elif vm['power_state'] == 'Paused':
+                    state = virt.Guest.STATE_PAUSED
+                elif vm['power_state'] == 'Halted':
+                    state = virt.Guest.STATE_SHUTOFF
+                else:
+                    state = virt.Guest.STATE_UNKNOWN
+
+                guests.append(virt.Guest(
+                        uuid= vm["uuid"],
+                        virt= self,
+                        state= state,
+                        hypervisorType=None
+                    )
+                )
+
+            mapping[record["hostname"]]=guests
```

Comment:
What does 'facts' mean? This would be the new code (but properly indented):

mapping['hypervisors']= Hypervisor (hypervisorId=hypervisor_record["uuid"], guestIds=guests, name=hypervisor_record["hostname"], facts="")